include/vcpkg/base/parallel-algorithms.h

```diff
@@ -1,85 +1,163 @@
 #pragma once
+#if defined(_WIN32)
+#include <vcpkg/base/system-headers.h>
+#else // ^^^ _WIN32 / !_WIN32 vvv
+#include <vcpkg/base/optional.h>
 
+#include <system_error>
+#include <thread>
+#include <type_traits>
+#include <vector>
+#endif // ^^^ _WIN32
 #include <vcpkg/base/system.h>
 
+#include <algorithm>
 #include <atomic>
-#include <future>
-#include <vector>
 
 namespace vcpkg
 {
     template<class F>
-    inline void execute_in_parallel(size_t work_count, F&& work) noexcept
+    struct WorkCallbackContext
     {
-        const size_t thread_count = static_cast<size_t>(get_concurrency());
-        const size_t num_threads = std::max(static_cast<size_t>(1), std::min(thread_count, work_count));
+        F work;
+        size_t work_count;
+        std::atomic<size_t> next_offset;
 
-        std::vector<std::future<void>> workers;
-        workers.reserve(num_threads - 1);
+        WorkCallbackContext(F init_f, size_t work_count) : work(init_f), work_count(work_count), next_offset(0) { }
```

Comment:
```suggestion
        WorkCallbackContext(F init_f, size_t work_count) : work(std::move(init_f)), work_count(work_count), next_offset(0) { }
```

Reply:
This is a functor/callable, which are usually pointers, and for which std::move is a pessimization. (This is also why they are passed by value)

Reply:
I understand why taking it by rvalue reference would be a pessimization, but how would std::move reduce performance?

Reply:
`std::move` forces that the value be bound to an rvalue-reference. This is why it, for example, breaks NRVO in `return std::move(blah);` cases. It is likely that it wouldn't matter in this case; I'm just used to 'pass the functors by value' resulting from the mess that caused us to add `_Pass_fn`: passing functors by reference broke the ability to inline through function pointers. Maybe my paranoia here is out of date by now but given that we control all the functors here and aren't going to ever pass ones that are expensive to copy I'm inclined to keep following the convention.

https://github.com/microsoft/STL/blob/bd3d740ae5de7255c720b8133c5d23aa131e0760/stl/inc/xutility#L558-L584